File: desktop/app/src/main/kotlin/com/skypc/core/Humanizer.kt
package com.skypc.core

import java.util.Random
import kotlin.math.cos
import kotlin.math.ln
import kotlin.math.sqrt

//按种子给每个音符生成稳定的拟人化误差
class Humanizer(private val hs: HumanizeSettings, private val seed: Int, private val count: Int) {
    private fun rnd(i: Int, salt: Int): Random = Random((seed xor (i * 486187739) xor (salt * 16777619)).toLong())

    private fun fatigue(i: Int): Double {
        if (!hs.fatigueEnabled) return 1.0
        val t = if (count <= 1) 0.0 else i.toDouble() / (count - 1)
        return hs.fatigueStart + (hs.fatigueEnd - hs.fatigueStart) * t
    }

    fun offset(i: Int): Double {
        if (!hs.jitterEnabled) return 0.0
        val lo = minOf(hs.jitterMin, hs.jitterMax).toDouble()
        val hi = maxOf(hs.jitterMin, hs.jitterMax).toDouble()
        val r = rnd(i, 1).nextDouble()
        val v = when (hs.jitterMode) {
            "Gaussian" -> {
                val mean = (lo + hi) / 2
                val sigma = (hi - lo) / 6
                (mean + sigma * gaussian(rnd(i, 7))).coerceIn(lo, hi)
            }
            "Early" -> lo + (hi - lo) * r * r
            "Late" -> hi - (hi - lo) * r * r
            "Triangular" -> lo + (hi - lo) * (r + rnd(i, 8).nextDouble()) / 2
            else -> lo + (hi - lo) * r
        }
        return v * hs.strength * fatigue(i)
    }

    fun hold(base: Int, i: Int): Int {
        if (!hs.holdEnabled) return base
        val lo = minOf(hs.holdMin, hs.holdMax).toDouble()
        val hi = maxOf(hs.holdMin, hs.holdMax).toDouble()
        val d = lo + (hi - lo) * rnd(i, 3).nextDouble()
        return maxOf(5, (base + d * hs.strength * fatigue(i)).toInt())
    }

    fun drop(i: Int): Boolean = hs.dropEnabled && rnd(i, 4).nextDouble() < hs.dropChance * hs.strength * fatigue(i)

    fun wrong(i: Int): Boolean = hs.wrongEnabled && rnd(i, 5).nextDouble() < hs.wrongChance * hs.strength * fatigue(i)

    private fun gaussian(r: Random): Double {
        val u1 = 1.0 - r.nextDouble()
        val u2 = 1.0 - r.nextDouble()
        return sqrt(-2.0 * ln(u1)) * cos(2.0 * Math.PI * u2)
    }
}

File: SkyPC_AutoMusic/Model/Humanizer.cs
using System;
using System.Collections.Generic;

namespace SkyPC_AutoMusic.Model
{
    //按当前设置给每一次演奏生成拟人化误差
    //同一个种子 + 同一个拍号，结果稳定可复现
    public class Humanizer
    {
        private readonly HumanizeSettings settings;
        private readonly int seed;
        private readonly int beatCount;
        private readonly double driftStart;
        private readonly double driftEnd;

        public Humanizer(HumanizeSettings settings, int seed, int beatCount)
        {
            this.settings = settings;
            this.seed = seed;
            this.beatCount = Math.Max(1, beatCount);

            Random pick = new Random(unchecked(seed ^ 0x5F3759DF));
            driftStart = Lerp(settings.DriftMin, settings.DriftMax, pick.NextDouble());
            driftEnd = Lerp(settings.DriftMin, settings.DriftMax, pick.NextDouble());
        }

        //每拍的触发时间偏移（毫秒）
        public double TriggerOffset(int beat, int timeMs)
        {
            double fatigue = Fatigue(beat);
            double jitter = settings.JitterEnabled ? SampleJitter(beat) * fatigue : 0;
            double drift = settings.DriftEnabled ? DriftOffset(beat, timeMs) : 0;
            return (jitter + drift) * settings.Strength;
        }

        //对当前拍音符做漏音/错音处理
        public List<NoteKey> ProcessKeys(List<NoteKey> keys, int beat)
        {
            if (!settings.DropEnabled && !settings.WrongEnabled)
                return keys;

            double factor = Fatigue(beat) * settings.Strength;
            Random rnd = Rnd(beat, 2);
            List<NoteKey> result = new List<NoteKey>(keys.Count);
            foreach (NoteKey key in keys)
            {
                if (settings.DropEnabled && rnd.NextDouble() < settings.DropChance * factor)
                    continue;
                NoteKey k = key;
                if (settings.WrongEnabled && rnd.NextDouble() < settings.WrongChance * factor)
                    k = Neighbor(k, rnd);
                result.Add(k);
            }
            return result;
        }

        //按键时长（毫秒）
        public int Hold(int beat, int baseMs)
        {
            if (!settings.HoldEnabled)
                return baseMs;

            double factor = Fatigue(beat) * settings.Strength;
            double min = Math.Min(settings.HoldMin, settings.HoldMax);
            double max = Math.Max(settings.HoldMin, settings.HoldMax);
            double delta = min + (max - min) * Rnd(beat, 3).NextDouble();
            return Math.Max(5, (int)Math.Round(baseMs + delta * factor));
        }

        //和弦各音之间的错峰间隔（毫秒），无启用则返回 null
        public int[] Spread(int keyCount, int beat)
        {
            if (!settings.SpreadEnabled || keyCount <= 1)
                return null;

            double min = Math.Min(settings.SpreadMin, settings.SpreadMax);
            double max = Math.Max(settings.SpreadMin, settings.SpreadMax);
            Random rnd = Rnd(beat, 4);
            int[] delays = new int[keyCount];
            for (int i = 0; i < keyCount; i++)
                delays[i] = Math.Max(0, (int)Math.Round((min + (max - min) * rnd.NextDouble()) * settings.Strength));
            return delays;
        }

        //空拍时可能插入的换气停顿（毫秒）
        public double Breath(int beat)
        {
            if (!settings.BreathEnabled)
                return 0;

            Random rnd = Rnd(beat, 5);
            if (rnd.NextDouble() > settings.BreathChance * settings.Strength)
                return 0;
            double min = Math.Min(settings.BreathMin, settings.BreathMax);
            double max = Math.Max(settings.BreathMin, settings.BreathMax);
            return (min + (max - min) * rnd.NextDouble()) * settings.Strength;
        }

        //极低概率插入一个装饰音
        public bool TryGrace(int beat, out NoteKey key)
        {
            key = NoteKey._1Key0;
            if (!settings.GraceEnabled)
                return false;
            Random rnd = Rnd(beat, 6);
            if (rnd.NextDouble() >= settings.GraceChance * settings.Strength)
                return false;
            key = (NoteKey)rnd.Next(0, 15);
            return true;
        }

        //从开头到结尾的疲劳倍率
        private double Fatigue(int beat)
        {
            if (!settings.FatigueEnabled)
                return 1.0;
            return Lerp(settings.FatigueStart, settings.FatigueEnd, beat / (double)beatCount);
        }

        private double DriftOffset(int beat, int timeMs)
        {
            double percent = Lerp(driftStart, driftEnd, beat / (double)beatCount);
            return percent / 100.0 * timeMs;
        }

        private double SampleJitter(int beat)
        {
            double min = Math.Min(settings.JitterMin, settings.JitterMax);
            double max = Math.Max(settings.JitterMin, settings.JitterMax);
            double range = max - min;
            Random rnd = Rnd(beat, 1);
            double r = rnd.NextDouble();

            switch (settings.JitterMode)
            {
                case "Gaussian":
                    double mean = (min + max) / 2.0;
                    double sigma = range / 6.0;
                    double g = mean + sigma * Gaussian(rnd);
                    return Math.Min(max, Math.Max(min, g));
                case "Triangular":
                    return min + range * (r + rnd.NextDouble()) / 2.0;
                case "Early":
                    return min + range * r * r;
                case "Late":
                    return max - range * r * r;
                case "Custom":
                    return SampleCustom(beat, min, max, r);
                default:
                    return min + range * r;
            }
        }

        private double SampleCustom(int beat, double min, double max, double r)
        {
            Dictionary<string, double> vars = new Dictionary<string, double>
            {
                { "r", r },
                { "min", min },
                { "max", max },
                { "beat", beat },
                { "count", beatCount },
                { "fatigue", Fatigue(beat) },
                { "strength", settings.Strength }
            };
            double value;
            if (HumanizeExpression.TryEval(settings.JitterExpr, vars, out value))
                return value;
            //表达式写错就退回均匀分布
            return min + (max - min) * r;
        }

        private static NoteKey Neighbor(NoteKey key, Random rnd)
        {
            int index = (int)key % 15;
            int step = rnd.Next(0, 2) == 0 ? -1 : 1;
            int next = index + step;
            if (next < 0) next = 1;
            if (next > 14) next = 13;
            return (NoteKey)next;
        }

        private static double Gaussian(Random rnd)
        {
            double u1 = 1.0 - rnd.NextDouble();
            double u2 = 1.0 - rnd.NextDouble();
            return Math.Sqrt(-2.0 * Math.Log(u1)) * Math.Cos(2.0 * Math.PI * u2);
        }

        private static double Lerp(double a, double b, double t)
        {
            return a + (b - a) * t;
        }

        //同种子、同拍号、同盐值 -> 同一个随机序列
        private Random Rnd(int beat, int salt)
        {
            unchecked
            {
                return new Random(seed + beat * 486187739 + salt * 16777619);
            }
        }
    }
}
